crates/linker/src/pe.rs

```diff
@@ -621,4 +751,169 @@ mod test {
+        let dir = tempfile::tempdir().unwrap();
+        let path = dir.path().join("test.exe");
+
+        let new_sections = [*b"placehol", *b"aaaabbbb"];
+
+        increase_number_of_sections_help(PE_DYNHOST, &new_sections, &path);
+    }
+
+    fn link_zig_host_and_app_help(dir: &Path) {
+        use object::ObjectSection;
+
+        let host_zig = indoc!(
+            r#"
+            const std = @import("std");
+
+            extern fn magic() callconv(.C) u64;
+
+            pub fn main() !void {
+                const stdout = std.io.getStdOut().writer();
+                try stdout.print("Hello, {}!\n", .{magic()});
+            }
+            "#
+        );
+
+        let app_zig = indoc!(
+            r#"
+            export fn magic() u64 {
+                return 42;
+            }
+            "#
+        );
+
+        let zig = std::env::var("ROC_ZIG").unwrap_or_else(|_| "zig".into());
+
+        std::fs::write(dir.join("host.zig"), host_zig.as_bytes()).unwrap();
+        std::fs::write(dir.join("app.zig"), app_zig.as_bytes()).unwrap();
+
+        let dylib_bytes = crate::generate_dylib::synthetic_dll(&["magic".into()]);
+        std::fs::write(dir.join("libapp.obj"), dylib_bytes).unwrap();
+
+        let output = std::process::Command::new(&zig)
+            .current_dir(dir)
+            .args(&[
+                "build-exe",
+                "libapp.obj",
+                "host.zig",
+                "-lc",
+                "-target",
+                "x86_64-windows-gnu",
+                "--strip",
+                "-OReleaseFast",
+            ])
+            .output()
+            .unwrap();
+
+        if !output.status.success() {
+            use std::io::Write;
+
+            std::io::stdout().write_all(&output.stdout).unwrap();
+            std::io::stderr().write_all(&output.stderr).unwrap();
+
+            panic!("zig build-exe failed");
+        }
+
+        let output = std::process::Command::new(&zig)
+            .current_dir(dir)
+            .args(&[
+                "build-obj",
+                "app.zig",
+                "-target",
+                "x86_64-windows-gnu",
+                "--strip",
+                "-OReleaseFast",
+            ])
+            .output()
+            .unwrap();
+
+        if !output.status.success() {
+            use std::io::Write;
+
+            std::io::stdout().write_all(&output.stdout).unwrap();
+            std::io::stderr().write_all(&output.stderr).unwrap();
+
+            panic!("zig build-obj failed");
+        }
+
+        let mut app = std::fs::read(dir.join("host.exe")).unwrap();
+        let dynamic_relocations = DynamicRelocationsPe::new(&app);
+
+        // Here we manually add the app.obj assembly to the host's .text section
+        // We can "just" do this because the space reserved for this section is bigger
+        // than what is actually used (because PE rounds up sections to a big alignment,
+        // here 0x200)
```

Comment:
Even if it's big alignment, couldn't you have a file that's just one byte short of the next 0x200, leaving you only one byte of room?

Reply:
yeah, this is a PoC. In the followup PR we actually make a custom extra code section with enough space to store all the app's machine code.